### PR TITLE
Avoid mutation of array-size expression in C++ `new` expression

### LINKED ISSUE #265: Dredd mutate the constant int used in aggregate initialisation
Dredd mutates the the sole statement in `main()` in the following file:
```
class a {
public:
  a(int) {}
};
int main() { 
    new a[2]{4, 5}; 
}
```
to
```
if (!__dredd_enabled_mutation(13)) { new a[__dredd_replace_expr_unsigned_long_constant(2, 0)]{__dredd_replace_expr_int_constant(4, 3), __dredd_replace_expr_int_constant(5, 8)}; }
```

This results in compilation error:
```
error: no matching constructor for initialization of 'a'
note: candidate constructor not viable: requires 1 argument, but 0 were provided
```

### PR DESCRIPTION
**Pull request by JonathanFoo0523**
In C++, one can initialize a list of items of type `a` with the expression `new a[2]{3, 4}`. Using a non-constant array size requires the type `a` to have a zero-argument constructor. Since applying mutation to the array size (i.e., `2` in the given example) would change the constant array size to non-constant, and we aren't sure whether type `a` contains a zero-argument constructor, we avoid mutating the array size altogether.

Fixes #265

### CODE REVIEW

File: test/single_file/new_expr_array_size.cc [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class foo {
+public:
+  foo(int x) {}
+};
+
+int main() { 
+  new foo[2]{4, 5}; 
+}
```
Comment (afd) [inline test/single_file/new_expr_array_size.cc:8]: Please there is a new line at end of file.

File: test/single_file/new_expr_array_size.cc.noopt.expected [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+class foo {
+public:
+  foo(int x) {}
+};
+
+int main() { 
+  if (!__dredd_enabled_mutation(12)) { new foo[2]{__dredd_replace_expr_int(4, 0), __dredd_replace_expr_int(5, 6)}; } 
+}
```
Comment (afd) [inline test/single_file/new_expr_array_size.cc.noopt.expected:63]: Same.

**Review by afd — APPROVED** [review]
This looks good - a couple of minor nits to do with new lines, but after those are fixed I'm happy for you to merge, so long as the CI bots pass.

When you merge, please do "squash and merge". Remove the PR number from the title. For the commit message, if your PR was multiple commits you'll get a bulleted list of messages by default - get rid of that, and just have one consolidated message. Wrap it at 72 chars (as per [this guidance](https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/)). And used #Fixes ... to that the related issue will be auto-closed.

**Comment by afd** [comment]
Your PR description, but wrapped to 72 chars, will make a good commit message. I'd change "Fix" to "Fixes" (can't remember if "Fix" will auto-close, but I prefer "Fixes" anyway).

File: test/single_file/new_expr_array_size.cc (hunk 1 above)
Comment (JonathanFoo0523) [reply test/single_file/new_expr_array_size.cc]: Fixed on [52f1fa6](https://github.com/mc-imperial/dredd/pull/272/commits/52f1fa68a33d9b0425c044b449ff5b4f7dbbae90)

File: test/single_file/new_expr_array_size.cc.noopt.expected (hunk 2 above)
Comment (JonathanFoo0523) [reply test/single_file/new_expr_array_size.cc.noopt.expected]: Fixed on [eaeec5d](https://github.com/mc-imperial/dredd/pull/272/commits/eaeec5d868e65d1134a68596db06850971bfe119)

**Pull request merged**